### ISSUE TITLE
Does ledger memory leak?

### ISSUE THREAD

**Issue by andrewshaoyu** [istio/istio#28687]
**Bug description**
when a istio resource (like serviceentry) updated 1qps,the pilot memory usage will increase quickly, and cpu util increased.
the resource update quickly is my problem,but why the memory usage keep increasing, does this situation meet the design?

[ ] Docs
[ ] Installation
[ ] Networking
[x] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**
memory need to be stable

**Steps to reproduce the bug**
``` go
	file, err := os.Create("mem-04.heap")
	if err != nil {
		fmt.Println(err)
		return
	}
	defer file.Close()
	lg := ledger.Make(time.Minute)
	for i := 0; i < 1000; i++ {
		_, err = lg.Put("a", strconv.Itoa(i))
		if err != nil {
			fmt.Println(err)
		}
	}
        time.Sleep(time.Minute * 2)
	pprof.WriteHeapProfile(file)
```
pprof the heap profile
``` sh
(pprof) top10
Showing nodes accounting for 1.50MB, 100% of 1.50MB total
      flat  flat%   sum%        cum   cum%
    1.50MB   100%   100%     1.50MB   100%  istio.io/pkg/ledger.(*smt).loadBatch
         0     0%   100%     1.50MB   100%  istio.io/pkg/ledger.(*smt).Update
         0     0%   100%     1.50MB   100%  istio.io/pkg/ledger.(*smt).loadChildren
         0     0%   100%     1.50MB   100%  istio.io/pkg/ledger.(*smt).update
         0     0%   100%     1.50MB   100%  istio.io/pkg/ledger.smtLedger.Put
         0     0%   100%     1.50MB   100%  main.main
         0     0%   100%     1.50MB   100%  runtime.main
ROUTINE ======================== istio.io/pkg/ledger.(*smt).loadBatch in /home/shaoyu/project/pkg/ledger/smt.go
    1.50MB     1.50MB (flat, cum)   100% of Total
         .          .    347:	s.db.updatedMux.RUnlock()
         .          .    348:	if exists {
         .          .    349:		if s.atomicUpdate {
         .          .    350:			// Return a copy so that Commit() doesnt have to be called at
         .          .    351:			// each block and still commit every state transition.
    1.50MB     1.50MB    352:			newVal := make([][]byte, batchLen)
         .          .    353:			copy(newVal, val)
         .          .    354:			return newVal, nil
         .          .    355:		}
         .          .    356:		return val, nil
         .          .    357:	}
```
**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**
istio 1.7.2
**How was Istio installed?**
istioctl generate manifests --set profile=default | kubectl apply -f
**Environment where bug was observed (cloud vendor, OS, etc)**
linux 4.19 kubernetes 1.17

**Comment by carolynhu** [comment]
https://github.com/istio/istio/issues/28015

**Comment by andrewshaoyu** [comment]
> https://github.com/istio/istio/issues/28015

there are different. pkg ledger it self Vs pkg ledger misused.

**Comment by andrewshaoyu** [comment]
Any progress?

**Comment by therealmitchconnors** [comment]
@andrewshaoyu the ledger does have a known slow leak in 1.7, but it shouldn't be enough to evict nodes (which seems to be in keeping with what you reported).  The ledger is being refactored in 1.9 to remove a few leak conditions, and add a features.  If I can get you a build of 1.9 which uses the refactored ledger, would you be willing to test it with the same configuration you used above?

**Comment by andrewshaoyu** [comment]
OK,I will try to test with 1.9.

**Comment by therealmitchconnors** [comment]
Fixed in #29275, currently cherrypicking into 1.9.
